In [1]:
import pandas as pd

In [2]:
flowers = pd.DataFrame({
    'color' : ['red', 'green', 'red', 'green', 'red', 'green', 'red', 'green', 'blue', 'blue'],
    'height': [4,9,4,8,4,7,4,7.5,20,19],
    'petals': [3,9,1,8,1,10,2,8,50,47],
    'days'  : [6,16,7,15,8,17,5,12,40,45]
})
flowers

color  height  petals  days
0    red     4.0       3     6
1  green     9.0       9    16
2    red     4.0       1     7
3  green     8.0       8    15
4    red     4.0       1     8
5  green     7.0      10    17
6    red     4.0       2     5
7  green     7.5       8    12
8   blue    20.0      50    40
9   blue    19.0      47    45

In [3]:
from sklearn.model_selection import train_test_split

In [4]:
X_train, X_test, y_train, y_test = train_test_split(flowers.drop('days', 1), flowers['days'],
                                                    test_size=.2, random_state=40)

C:\Users\ps109\AppData\Local\Temp\ipykernel_17284\3169547743.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X_train, X_test, y_train, y_test = train_test_split(flowers.drop('days', 1), flowers['days'],


In [5]:
from sklearn.preprocessing import OneHotEncoder

In [6]:
ohe = OneHotEncoder(sparse=False)

In [7]:
X_train[['color']]

color
8   blue
1  green
2    red
9   blue
0    red
5  green
7  green
6    red

In [10]:
X_train['color'].unique()

array(['blue', 'green', 'red'], dtype=object)

In [16]:
arr1=X_train['color'].to_numpy()

In [20]:
for i in range(len(arr1)):
    if arr1[i] == 'blue':
        print(i)
        break
        

0


In [21]:
for i in range(len(arr1)):
    if arr1[i] == 'green':
        print(i)
        break

1


In [8]:
ohe.fit(X_train[['color']])

OneHotEncoder(sparse=False)

In [9]:
ohe.transform(X_train[['color']])

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.],
       [1., 0., 0.],
       [0., 0., 1.],
       [0., 1., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [10]:
X_test

color  height  petals
4    red     4.0       1
3  green     8.0       8

In [11]:
ohe.transform(X_test[['color']])

array([[0., 0., 1.],
       [0., 1., 0.]])

In [12]:
flowers_new = pd.DataFrame({
    'color' : ['red', 'green', 'red', 'pink', 'red', 'green', 'red', 'green', 'blue', 'blue'],
    'height': [4,9,4,8,4,7,4,7.5,20,19],
    'petals': [3,9,1,8,1,10,2,8,50,47],
    'days'  : [6,16,7,15,8,17,5,12,40,45]
})
flowers_new

color  height  petals  days
0    red     4.0       3     6
1  green     9.0       9    16
2    red     4.0       1     7
3   pink     8.0       8    15
4    red     4.0       1     8
5  green     7.0      10    17
6    red     4.0       2     5
7  green     7.5       8    12
8   blue    20.0      50    40
9   blue    19.0      47    45

In [13]:
X_1, X_2, y_1, y_2 = train_test_split(flowers_new.drop('days', 1), flowers_new['days'],
                                                    test_size=.2, random_state=40)

C:\Users\91760\AppData\Local\Temp/ipykernel_9484/4158383079.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  X_1, X_2, y_1, y_2 = train_test_split(flowers_new.drop('days', 1), flowers_new['days'],


In [14]:
enc1 = OneHotEncoder(sparse=False, handle_unknown='ignore')

In [15]:
X_1['color']

8     blue
1    green
2      red
9     blue
0      red
5    green
7    green
6      red
Name: color, dtype: object

In [16]:
X_2

color  height  petals
4   red     4.0       1
3  pink     8.0       8

In [17]:
enc1.fit_transform(X_1[['color']])

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.],
       [1., 0., 0.],
       [0., 0., 1.],
       [0., 1., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [18]:
enc1.transform(X_2[['color']])

array([[0., 0., 1.],
       [0., 0., 0.]])

In [19]:
df = pd.read_csv('data/income_evaluation.csv')
df.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country income  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

In [20]:
X_train, X_test, y_train, y_test = train_test_split(df.drop('income', 1), df[' income'],
                                                    test_size=.2, random_state=40)

C:\Users\91760\AppData\Local\Temp/ipykernel_9484/73320746.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  X_train, X_test, y_train, y_test = train_test_split(df.drop(' income', 1), df[' income'],


KeyError: "[' income'] not found in axis"

In [ ]:
X_train[' workclass'].unique()

In [22]:
o1 = OneHotEncoder()

In [23]:
o1.fit_transform(X_train[[' workclass']]).toarray()

array([[0., 0., 0., ..., 0., 0., 0.],
       [1., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [1., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 1., 0., 0.]])

In [24]:
pd.DataFrame(o1.fit_transform(X_train[[' workclass']]).toarray())

0    1    2    3    4    5    6    7    8
0      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
1      1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
2      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
3      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
4      1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
5      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
6      1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
7      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
8      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
9      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
10     0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0
11     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
12     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
13     0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0
14     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
15     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
16     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
17     0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
18     0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0
19     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
20     0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
21     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
22     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
23     0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0
24     0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0
25     0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
26     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
27     0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
28     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
29     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
...    ...  ...  ...  ...  ...  ...  ...  ...  ...
26018  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26019  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26020  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26021  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26022  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26023  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26024  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26025  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26026  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26027  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
26028  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
26029  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26030  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26031  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
26032  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
26033  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26034  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26035  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
26036  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26037  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26038  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26039  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
26040  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
26041  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26042  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26043  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26044  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0
26045  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
26046  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
26047  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0

[26048 rows x 9 columns]

In [25]:
o1.transform(X_test[[' workclass']]).toarray()

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [1., 0., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [26]:
o2 = OneHotEncoder(sparse=False)

In [27]:
o2.fit_transform(X_train[[' workclass', ' occupation']])

array([[0., 0., 0., ..., 0., 0., 0.],
       [1., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 1., 0., 0.],
       ...,
       [1., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [28]:
X_train[' workclass'].nunique()

9

In [29]:
pd.DataFrame(o2.fit_transform(X_train[[' workclass', ' occupation']]))

0    1    2    3    4    5    6    7    8    9   ...   14   15   16  \
0      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1      1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0   
2      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
3      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  1.0  0.0   
4      1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0   
5      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  1.0  0.0   
6      1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0   
7      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
8      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
9      0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
10     0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
11     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
12     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
13     0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...  0.0  0.0  0.0   
14     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
15     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
16     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
17     0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
18     0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
19     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
20     0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
21     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
22     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
23     0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
24     0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...  0.0  0.0  0.0   
25     0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  1.0  0.0  0.0   
26     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
27     0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
28     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
29     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
26018  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
26019  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
26020  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  1.0   
26021  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  1.0   
26022  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
26023  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
26024  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  1.0   
26025  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  1.0  0.0   
26026  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
26027  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
26028  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0   
26029  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  1.0   
26030  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
26031  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0   
26032  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0   
26033  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
26034  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  1.0   
26035  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0   
26036  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  .

In [30]:
o2.get_feature_names()

array(['x0_ ?', 'x0_ Federal-gov', 'x0_ Local-gov', 'x0_ Never-worked',
       'x0_ Private', 'x0_ Self-emp-inc', 'x0_ Self-emp-not-inc',
       'x0_ State-gov', 'x0_ Without-pay', 'x1_ ?', 'x1_ Adm-clerical',
       'x1_ Armed-Forces', 'x1_ Craft-repair', 'x1_ Exec-managerial',
       'x1_ Farming-fishing', 'x1_ Handlers-cleaners',
       'x1_ Machine-op-inspct', 'x1_ Other-service',
       'x1_ Priv-house-serv', 'x1_ Prof-specialty', 'x1_ Protective-serv',
       'x1_ Sales', 'x1_ Tech-support', 'x1_ Transport-moving'],
      dtype=object)

In [31]:
pd.set_option('display.max_columns', None)

In [32]:
pd.DataFrame(o2.fit_transform(X_train[[' workclass', ' occupation']]), columns=o2.get_feature_names())

x0_ ?  x0_ Federal-gov  x0_ Local-gov  x0_ Never-worked  x0_ Private  \
0        0.0              0.0            0.0               0.0          1.0   
1        1.0              0.0            0.0               0.0          0.0   
2        0.0              0.0            0.0               0.0          1.0   
3        0.0              0.0            0.0               0.0          1.0   
4        1.0              0.0            0.0               0.0          0.0   
5        0.0              0.0            0.0               0.0          1.0   
6        1.0              0.0            0.0               0.0          0.0   
7        0.0              0.0            0.0               0.0          1.0   
8        0.0              0.0            0.0               0.0          1.0   
9        0.0              0.0            0.0               0.0          1.0   
10       0.0              0.0            1.0               0.0          0.0   
11       0.0              0.0            0.0               0.0          1.0   
12       0.0              0.0            0.0               0.0          1.0   
13       0.0              0.0            0.0               0.0          0.0   
14       0.0              0.0            0.0               0.0          1.0   
15       0.0              0.0            0.0               0.0          1.0   
16       0.0              0.0            0.0               0.0          1.0   
17       0.0              0.0            0.0               0.0          0.0   
18       0.0              0.0            0.0               0.0          0.0   
19       0.0              0.0            0.0               0.0          1.0   
20       0.0              0.0            0.0               0.0          0.0   
21       0.0              0.0            0.0               0.0          1.0   
22       0.0              0.0            0.0               0.0          1.0   
23       0.0              0.0            1.0               0.0          0.0   
24       0.0              0.0            0.0               0.0          0.0   
25       0.0              0.0            0.0               0.0          0.0   
26       0.0              0.0            0.0               0.0          1.0   
27       0.0              0.0            0.0               0.0          0.0   
28       0.0              0.0            0.0               0.0          1.0   
29       0.0              0.0            0.0               0.0          1.0   
...      ...              ...            ...               ...          ...   
26018    0.0              0.0            0.0               0.0          1.0   
26019    0.0              0.0            0.0               0.0          1.0   
26020    0.0              0.0            0.0               0.0          1.0   
26021    0.0              0.0            0.0               0.0          1.0   
26022    0.0              0.0            0.0               0.0          1.0   
26023    0.0              0.0            0.0               0.0          1.0   
26024    0.0              0.0            0.0               0.0          1.0   
26025    0.0              0.0            0.0               0.0          1.0   
26026    0.0              0.0            0.0               0.0          1.0   
26027    0.0              0.0            0.0               0.0          0.0   
26028    1.0              0.0            0.0               0.0          0.0   
26029    0.0              0.0            0.0               0.0          1.0   
26030    0.0              0.0            0.0               0.0          1.0   
26031    1.0              0.0            0.0               0.0          0.0   
26032    1.0              0.0            0.0               0.0          0.0   
26033    0.0              0.0            0.0               0.0          1.0   
26034    0.0              0.0            0.0               0.0          1.0   
26035    1.0              0.0            0.0               0.0          0.0   
26036    0.0              0.0            0.0             

In [33]:
categ = [col for col in X_train.columns if X_train[col].dtypes == 'O']

In [34]:
o3 = OneHotEncoder(sparse=False)

In [35]:
o3.fit_transform(X_train[categ])

array([[0., 0., 0., ..., 0., 0., 0.],
       [1., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 1., 0., 0.],
       ...,
       [1., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 1., 0., 0.]])

In [36]:
X_train[' workclass'].unique()

array([' Private', ' ?', ' Local-gov', ' State-gov', ' Self-emp-not-inc',
       ' Self-emp-inc', ' Federal-gov', ' Never-worked', ' Without-pay'],
      dtype=object)

In [37]:
pd.DataFrame(o3.fit_transform(X_train[categ])).shape

(26048, 102)

In [38]:
len(categ)

8

## now using drop='first', we'll notice how 8 columns have been removed, and now we have 102 - 8 = 94 columns.

In [39]:
o4 = OneHotEncoder(sparse=False, drop='first')

In [40]:
o4.fit_transform(X_train[categ])

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 1., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 1., 0., 0.]])

In [41]:
pd.DataFrame(o4.fit_transform(X_train[categ])).shape

(26048, 94)